File: epi_judge_cpp/search_entry_equal_to_index.cc
#include <vector>

#include "test_framework/test_failure_exception.h"
#include "test_framework/test_timer.h"

using std::vector;

int SearchEntryEqualToItsIndex(const vector<int>& A) {
  // Implement this placeholder.
  return 0;
}

void SearchEntryEqualToItsIndexWrapper(TestTimer& timer, const vector<int>& A) {
  timer.Start();
  int result = SearchEntryEqualToItsIndex(A);
  timer.Stop();
  if (result != -1) {
    if (A[result] != result) {
      throw TestFailureException("Entry does not equal to its index");
    }
  } else {
    for (int i = 0; i < A.size(); ++i) {
      if (A[i] == i) {
        throw TestFailureException(
            "There are entries which equal to its index");
      }
    }
  }
}

#include "test_framework/generic_test.h"

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"timer", "A"};
  GenericTestMain(args, "search_entry_equal_to_index.tsv",
                  &SearchEntryEqualToItsIndexWrapper, DefaultComparator{},
                  param_names);
  return 0;
}

File: epi_judge_java/epi/SearchEntryEqualToIndex.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;
import epi.test_framework.TimedExecutor;

import java.util.List;

public class SearchEntryEqualToIndex {

  public static int searchEntryEqualToItsIndex(List<Integer> A) {
    // Implement this placeholder.
    return 0;
  }

  @EpiTest(testfile = "search_entry_equal_to_index.tsv")
  public static void searchEntryEqualToItsIndexWrapper(TimedExecutor executor,
                                                       List<Integer> A)
      throws Exception {
    int result = executor.run(() -> searchEntryEqualToItsIndex(A));

    if (result != -1) {
      if (A.get(result) != result) {
        throw new TestFailure("Entry does not equal to its index");
      }
    } else {
      for (int i = 0; i < A.size(); ++i) {
        if (A.get(i) == i) {
          throw new TestFailure("There are entries which equal to its index");
        }
      }
    }
  }

  public static void main(String[] args) {
    // The timeout is set to 15 seconds for each test case.
    // If your program ends with TIMEOUT error, and you want to try longer time
    // limit, you can extend the limit by changing the following line.
    long timeoutSeconds = 15;

    System.exit(
        GenericTest
            .runFromAnnotations(args, timeoutSeconds,
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
